### DISCUSSION TITLE
Task Covariance Prior in KroneckerMultiTaskGP model

### DISCUSSION THREAD

**Discussion by higherorderderivatives2024** [meta-pytorch/botorch#1560] [Q&A]
In the KroneckerMultiTaskGP [model](https://botorch.org/api/_modules/botorch/models/multitask.html#FixedNoiseMultiTaskGP), one of the inputs is the prior on the task covariance matrix. Which of the following is true?:

1. When the KroneckerMultiTaskGP model is initialized, a task covariance matrix is sampled from the given prior, and held fixed throughout, including when sampling from model.posterior.
2. Calling model.posterior automatically devises the means to sample from the posterior of the task covariance matrix, given the prior and the data. Hence, calling model.posterior learns the task covariance matrix from the data. 

Secondly, KroneckerMultiTaskGP uses a gamma prior on the noise. Is this the sigma^2 value when we add a '\sigma^2 \mathbf I_{nt}' term in the noise model for the global noise? Like earlier, does calling model.posterior automatically devise the means to sample from the noise posterior, given the prior and the data? (Hence, does calling model.posterior learn the noise from the data?). Or is the \sigma^2 value sampled once and held fixed?

Thanks!

**Comment by Balandat** [comment]
The model automatically learns the task covariance matrix from the data. The same is true for the noise level. 

(note that while these learned values are used when calling `posterior`, they are inferred during the model fitting)

**Reply by higherorderderivatives2024** [reply]
I see, thanks! So, when the function fit_gpytorch_mll(mll) (which i understand stands for maximum likelihood estimation) is called, 

1. Since we are specifying a likelihood in mll, are the task covariance matrix and noise level learned using Maximum Likelihood Estimation when fitting the model? Or,
2. Since a prior is given over those values, is a posterior over those values learned during model fitting, when is then sampled from when calling model.posterior?

Are there any resources you could recommend to understand the underlying computations that occur in noise/covariance matrix fitting?

**Reply by Balandat** [reply]
Yes, task covariance matrix and noise level are learned via MLE, incorporating the provided priors in the likelihood. 

> Are there any resources you could recommend to understand the underlying computations that occur in noise/covariance matrix fitting?

At an abstract level, section 5.4.1 of [1] explains what's going on with computing the MLL, only that you'll need to add the additional likelihood terms of the priors over task covariance and noise, i.e. sth like `p(task_covar | theta) p(sigma^2 | theta)`. 

In code this is not particularly explicit, but the terms are added in the `_add_other_terms` call [here](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/mlls/exact_marginal_log_likelihood.py#L62-L65) that looks up priors registered on the module level and computes these (log) likelihoods. 

[1] C. E. Rasmussen and C. K. Williams. Gaussian processes for machine learning. 2006. The MIT Press, Cambridge, MA, USA, 38:715–719, 2006.

**Reply by higherorderderivatives2024** [reply]
Thanks! 

From Section 5.1, it seems that the theta vector in the text includes sigma^2. Since we have the prior for sigma^2, by "MLE which incorporates the given priors" do you mean that fit_gpytorch_mll(mll) finds the [Maximum Apriori Posterior](https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation) estimate for sigma^2 and the task covariance matrix (i.e. the maximum of the posterior distribution for theta and the covariance matrix given the prior and likelihood?)

**Reply by Balandat** [reply]
You can find how the kernel works here: https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/kernels/multitask_kernel.py#L13. The parameters that are fit are contained in the  `IndexKernel`: https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/kernels/index_kernel.py#L18

**Comment by Minssnail** [comment]
收到啦！我会尽快回复您。
